react/features/base/conference/middleware.web.ts

```diff
@@ -19,15 +19,68 @@ let screenLock: WakeLockSentinel | undefined;
 /**
  * Releases the screen lock.
  *
- * @returns {void}
+ * @returns {Promise}
  */
-function releaseScreenLock() {
+async function releaseScreenLock() {
     if (screenLock) {
-        screenLock.release();
+        if (!screenLock.released) {
+            logger.debug('Releasing wake lock.');
+
+            try {
+                await screenLock.release();
+            } catch (e) {
+                logger.error(`Error while releasing the screen wake lock: ${e}.`);
+            }
+        }
+        screenLock.removeEventListener('release', onWakeLockReleased);
         screenLock = undefined;
+        document.removeEventListener('visibilitychange', handleVisibilityChange);
     }
 }
 
+/**
+ * Requests a new screen wake lock.
+ *
+ * @returns {void}
+ */
+function requestWakeLock() {
+    if (navigator.wakeLock?.request) {
+        navigator.wakeLock.request('screen')
+            .then(lock => {
+                screenLock = lock;
+                screenLock.addEventListener('release', onWakeLockReleased);
+                document.addEventListener('visibilitychange', handleVisibilityChange);
+                logger.debug('Wake lock created.');
+            })
+            .catch(e => {
+                logger.error(`Error while requesting wake lock for screen: ${e}`);
+            });
+    }
+}
+
+/**
+ * Page visibility change handler that re-requests the wake lock if it has been released by the OS.
+ *
+ * @returns {void}
+ */
+function handleVisibilityChange() {
+    if (screenLock?.released && document.visibilityState === 'visible') {
+        // The screen lock have been released by the OS because of document visibility change. Lets try to request the
+        // wake lock again.
+        releaseScreenLock();
```

Comment:
Shouldn't you here wait for the release to finish before requesting again?

Reply:
In practice not really because we will be already released by the OS and we will just remove some listeners which should be synchronous.

Otherwise in theory I see how this can be more intuitive and I can change the `releaseScreenLock()` function to return Promise and do `await` here. Doing this will add extra logic and we will get a slight performance hit because of the promise, that's why my personal preference will be to leave it as it is, but if you think it will be clearer I will change it? WDYT?

Reply:
You have no guarantee that the logic in release will execute before the logic in request, and release if executed after the request can clean up the listeners that were added from requestWakeLock, isn't it?

Reply:
I have - `removeEventListener` is synchronous. Therefore we will first remove the listeners and then call `requestWakeLock`.

Reply:
But nothing waits remove to be executed. It can be scheduled and executed after the add of the listeners.

Reply:
@hristoterezov The performance argument is moot, this is not a hot path.

I agree with @damencho that it's clearer if we await because the sequence of actions is an Josu without having to go through the code.

Reply:
@damencho let's discuss this offline. I'm having a hard time understanding what you mean.

@saghul I agree this is not a hot path. Anyway I tend to try to not add redundant promises. In reality in this logic we don't care when exactly the screen lock is released. We care that we executed release. Isn't it? Am I missing something?

Reply:
Got it. The function is async and it will return promise anyway. Will add an await.